### ISSUE TITLE
Feature request: Switch for DataIterators whether batches smaller than batch_size should be skipped

### ISSUE THREAD

**Issue by mnschmit** [allenai/allennlp#3118]
In scenarios where the number of samples are not dividable by the batch size, the last batch cannot be filled completely.
If one needs fixed-length batches, one might want to discard those "incomplete" batches during training.

I haven't been able to find such a switch in the documentation although it seems to be a rather simple feature.

Use cases are scenarios where one needs a guaranteed minimal batch size, such as batch normalization where batches should always be bigger than 1, or models that use a buffer depending on a fixed batch size.

Usage example:
```
buck_iter = BucketIterator([("input", "num_tokens")], skip_smaller_batches=True)
```

**Comment by kernelmachine** [comment]
I think this is a reasonable feature request. Perhaps @joelgrus or @brendan-ai2 have some additional ideas on implementation details. In general, contributions welcome!

**Comment by mnschmit** [comment]
Thank you for your reply @kernelmachine !
I have an ad-hoc solution that just checks for the length in `BucketIterator`'s `_create_batches`:
```
batches = []
excess: Deque[Instance] = deque()
for batch_instances in lazy_groups_of(iter(instance_list), self._batch_size):
    for possibly_smaller_batches in self._ensure_batch_is_sufficiently_small(batch_instances, excess):
        if self._skip_smaller_batches and len(possibly_smaller_batches) < self._batch_size:
            continue
        batches.append(Batch(possibly_smaller_batches))
if excess and (not self._skip_smaller_batches or len(excess) == self._batch_size):
    batches.append(Batch(excess))
```

I have two problems with that solution:
1. It feels like this is suboptimal. Probably it would be better if `possibly_smaller_batches` could not be 'possibly smaller' in the first place instead of checking the length afterwards.
2. The change should probably be in the `DataIterator` class. So all iterators could benefit from it. Do you have any recommendation where to put the length check? Should it be in `_ensure_batch_is_sufficiently_small`? I suppose the `excess` will always have to be checked in the inheriting class though.

I'll be happy to contribute my changes if we agree on a good solution.

**Comment by DeNeutoy** [comment]
This is important for models with batchnorm, so we'll leave this as contributions welcome for the bucket iterator.
